server/src/main/java/com/thoughtworks/go/server/service/RulesService.java

```diff
@@ -41,18 +46,103 @@ public RulesService(GoConfigService goConfigService) {
+                .forEach(secretConfig -> {
+                    pipelines.forEach(pipelineName -> {
+                        MaterialConfig materialConfig = goConfigService
+                                .findPipelineByName(pipelineName)
+                                .materialConfigs()
+                                .getByMaterialFingerPrint(scmMaterial.getFingerprint());
+                        if (materialConfig != null) {
+                            PipelineConfigs group = goConfigService.findGroupByPipeline(pipelineName);
+                            if (!secretConfig.canRefer(group.getClass(), group.getGroup())) {
+                                if (!pipelinesWithViolatedSecretConfigIds.containsKey(pipelineName)) {
+                                    pipelinesWithViolatedSecretConfigIds.put(pipelineName, new ArrayList<>());
+                                }
+                                String errorMessage = format("Pipeline '%s' does not have permission to refer to secrets using SecretConfig: '%s'", pipelineName.toString(), secretConfig.getId());
+                                LOGGER.error("[Material Update] Failure: {}", errorMessage);
+                                pipelinesWithViolatedSecretConfigIds.get(pipelineName).add(secretConfig.getId());
+
+                            }
+                        }
+                    });
+                });
+
+        if (pipelinesWithViolatedSecretConfigIds.size() == pipelines.size()) {
+            String errorMessageFormat = "Pipeline(s) '%s' are using the material with URL: '%s', and do not have permissions to refer secret config '%s'.";
+            StringBuilder errorMessage = new StringBuilder();
+            Iterator<Map.Entry<CaseInsensitiveString, List<String>>> iterator = pipelinesWithViolatedSecretConfigIds.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<CaseInsensitiveString, List<String>> entry = iterator.next();
+                errorMessage.append(format(errorMessageFormat, entry.getKey().toString(), scmMaterial.getUriForDisplay(), entry.getValue().toString()));
+            }
+            LOGGER.error("[Material Update] Failure: {}", errorMessage.toString());
+            throw new RulesViolationException(errorMessage.toString());
+        }
         return true;
     }
 
+    public boolean validateSecretConfigReferences(ScmMaterial scmMaterial) {
+        List<CaseInsensitiveString> pipelines = goConfigService.pipelinesWithMaterial(scmMaterial.getFingerprint());
+
+        HashMap<CaseInsensitiveString, StringBuilder> pipelinesWithErrors = new HashMap<>();
+        pipelines.forEach(pipelineName -> {
+            MaterialConfig materialConfig = goConfigService
+                    .findPipelineByName(pipelineName)
+                    .materialConfigs()
+                    .getByMaterialFingerPrint(scmMaterial.getFingerprint());
+            if (materialConfig instanceof ScmMaterialConfig) {
```

Comment:
I don't think this is required as the method is only taking ScmMaterial as an input